Assets/MixedRealityToolkit/_Core/Definitions/InteractionDefinition.cs

```diff
@@ -9,19 +12,178 @@ namespace Microsoft.MixedReality.Toolkit.Internal.Definitions
     /// </summary>
     public struct InteractionDefinition
     {
+        public InteractionDefinition(uint id, AxisType axisType, InputType inputType) : this()
+        {
+            Id = id;
+            AxisType = axisType;
+            InputType = inputType;
+        }
+
+        #region Interaction Properties
+
         /// <summary>
-        /// The ID assigned to the Input
+        /// The Id assigned to the Interaction.
         /// </summary>
-        public string Id { get; set; }
+        public uint Id { get; }
 
         /// <summary>
         /// The axis type of the button, e.g. Analogue, Digital, etc.
         /// </summary>
-        public AxisType AxisType { get; set; }
+        public AxisType AxisType { get; }
+
+        /// <summary>
+        /// The primary action of the input as defined by the controller SDK.
+        /// </summary>
+        public InputType InputType { get; }
+
+        /// <summary>
+        /// Has the value changed since the last reading.
+        /// </summary>
+        public bool Changed { get; private set; }
+
+        #endregion Interaction Properties
+
+        #region Definition Data items
+
+        /// <summary>
+        /// The data storage for a Raw / None Axis type.
+        /// </summary>
+        private object RawData { get; set; }
+
+        /// <summary>
+        /// The data storage for a Digital Axis type.
+        /// </summary>
+        private bool BoolData { get; set; }
+
+        /// <summary>
+        /// The data storage for a Single Axis type.
+        /// </summary>
+        private float FloatData { get; set; }
+
+        /// <summary>
+        /// The data storage for a Dual Axis type.
+        /// </summary>
+        private Vector2 Vector2Data { get; set; }
 
         /// <summary>
-        /// The primary action of the button as defined by the controller SDK.
+        /// The position data storage for a 3DoF or 6DoF Axis type.
         /// </summary>
-        public InputType InputType { get; set; }
+        private Vector3 PositionData { get; set; }
+
+        /// <summary>
+        /// The rotation data storage for a 3DoF or 6DoF Axis type.
+        /// </summary>
+        private Quaternion RotationData { get; set; }
+
+        private Tuple<Vector3, Quaternion> TransformData { get; set; }
+
+        #endregion Definition Data items
+
+        #region Get Operators
+
+        public object GetRaw()
```

Comment:
As discussed, I personally don't like the multiple GET option as it complicates the API.

We should have the following allowing the model to store and retrieve the data as it sees fit:
* Set Data
* Get Data


Where as this is now:


* Set Data

* Get Bool

* Get Float

* Get....


Which makes the code look very messy.  I'd prefer to keep the original model I laid out and only return to this if we hit issues.

Reply:
Just wanted to avoid boxing

Reply:
Can we do both?